libraries/AP_Servo_Telem/AP_Servo_Telem.cpp

```diff
@@ -0,0 +1,160 @@
+{
+    return (last_update_ms == 0) || ((now_ms - last_update_ms) > 5000);
+}
+
+// return true if the requested types of data are available
+bool AP_Servo_Telem::TelemetryData::present(const uint16_t type_mask) const volatile
+{
+    return (valid_types & type_mask) != 0;
+}
+
+// return true if the requested types of data are available and not stale
+bool AP_Servo_Telem::TelemetryData::valid(const uint16_t type_mask) const volatile
+{
+    return present(type_mask) && !stale(AP_HAL::millis());
+}
+
+// record an update to the telemetry data together with timestamp
+// callback to update the data in the frontend, should be called by the driver when new data is available
+void AP_Servo_Telem::update_telem_data(const uint8_t servo_index, const TelemetryData& new_data)
+{
+    // telemetry data are not protected by a semaphore even though updated from different threads
+    // each element is a primitive type and the timestamp is only updated at the end, thus a caller
+    // can only get slightly more up-to-date information that perhaps they were expecting or might
+    // read data that has just gone stale - both of these are safe and avoid the overhead of locking
+
+    if (servo_index >= ARRAY_SIZE(_telem_data) || (new_data.valid_types == 0)) {
+        return;
+    }
+
+    volatile TelemetryData &telemdata = _telem_data[servo_index];
+
+    if (new_data.present(TelemetryData::Types::COMMANDED_POSITION)) {
+        telemdata.command_position = new_data.command_position;
+    }
+    if (new_data.present(TelemetryData::Types::MEASURED_POSITION)) {
+        telemdata.measured_position = new_data.measured_position;
+    }
+    if (new_data.present(TelemetryData::Types::FORCE)) {
+        telemdata.force = new_data.force;
+    }
+    if (new_data.present(TelemetryData::Types::SPEED)) {
+        telemdata.speed = new_data.speed;
+    }
+    if (new_data.present(TelemetryData::Types::VOLTAGE)) {
+        telemdata.voltage = new_data.voltage;
+    }
+    if (new_data.present(TelemetryData::Types::CURRENT)) {
+        telemdata.current = new_data.current;
+    }
+    if (new_data.present(TelemetryData::Types::DUTY_CYCLE)) {
+        telemdata.duty_cycle = new_data.duty_cycle;
+    }
+    if (new_data.present(TelemetryData::Types::MOTOR_TEMP)) {
+        telemdata.motor_temperature_cdeg = new_data.motor_temperature_cdeg;
+    }
+    if (new_data.present(TelemetryData::Types::PCB_TEMP)) {
+        telemdata.pcb_temperature_cdeg = new_data.pcb_temperature_cdeg;
+    }
+    if (new_data.present(TelemetryData::Types::PCB_TEMP)) {
+        telemdata.status_flags = new_data.status_flags;
+    }
+
+    telemdata.valid_types |= new_data.valid_types;
+    telemdata.last_update_ms = AP_HAL::millis();
+}
+
+void AP_Servo_Telem::update()
+{
+#if HAL_LOGGING_ENABLED
+    write_log();
+#endif
+}
+
+#if HAL_LOGGING_ENABLED
+void AP_Servo_Telem::write_log()
+{
+    AP_Logger *logger = AP_Logger::get_singleton();
```

Comment:
we should have an active_mask and return immediately for the most common case of no actuators active